AudioKit/Common/Nodes/Playback/Player/AKPlayer.swift

```diff
@@ -0,0 +1,613 @@
+//
+
+import AVFoundation
+
+/**
+ AKPlayer is meant to be a simple yet powerful audio player that just works. It supports
+ scheduling of sounds, looping, fading, and reversing. Players can be locked to a common
+ clock as well as video by using hostTime in the various play functions.
+ 
+ A locked video function would resemble:
+ ```
+ func videoPlay(at time: TimeInterval = 0, hostTime: UInt64 = 0 ) {
+     let cmHostTime = CMClockMakeHostTimeFromSystemUnits(hostTime)
+     let cmVTime = CMTimeMakeWithSeconds(time, 1000000)
+     let futureTime = CMTimeAdd(cmHostTime, cmVTime)
+     videoPlayer.setRate(1, time: kCMTimeInvalid, atHostTime: futureTime)
+ }
+ ```
+ 
+ Basic usage looks like:
+ ```
+    guard let player = AKPlayer(url: url) else { return }
+    player.completionHandler = { Swift.print("Done") }
+
+    // Loop Options
+    player.loop.start = 1
+    player.loop.end = 3
+    player.isLooping = true
+ 
+    player.play()
+ ```
+ 
+ Please note that pre macOS 10.13 / iOS 11 the completionHandler isn't sample accurate. It's pretty close though.
+ */
+public class AKPlayer: AKNode {
+    
+    /// How the player should handle audio. If buffering, it will load the audio data into
+    /// an internal buffer and play from ram. If not, it will play the file from disk.
+    /// Dynamic buffering will only load the audio if it needs to for processing reasons.
+    public enum BufferingType {
+        case dynamic, always, never
+    }
+    
+    //TODO: allow for different exponential curve slopes, implement other types
+    public enum FadeType {
+        case exponential //, linear, logarithmic
+    }
+    
+    public struct Loop {
+        var start: Double = 0
+        var end: Double = 0
+    }
+    
+    public struct Fade {
+        static var defaultStartGain: Double = 0.001
+        
+        var needsUpdate: Bool = false
+        
+        var inTime: Double = 0 {
+            willSet {
+                if newValue != inTime { needsUpdate = true }
+            }
+        }
+        
+        var inStartGain: Double = defaultStartGain {
+            willSet {
+                if newValue != inStartGain { needsUpdate = true }
+            }
+        }
+        
+        var outTime: Double = 0 {
+            willSet {
+                if newValue != outTime { needsUpdate = true }
+            }
+        }
+        
+        var outStartGain: Double = 1 {
+            willSet {
+                if newValue != outStartGain { needsUpdate = true }
+            }
+        }
+        
+        var type: AKPlayer.FadeType = .exponential {
+            willSet {
+                if newValue != type { needsUpdate = true }
+            }
+        }
+    }
+    
+    //MARK:- Private Parts
+    
+    // The underlying player node
+    private let playerNode = AVAudioPlayerNode()
+    private var mixer = AVAudioMixerNode()
+    private var buffer: AVAudioPCMBuffer?
+    private var startingFrame: AVAudioFramePosition = -1
```

Comment:
Why are these -1?

Reply:
let updateNeeded = (buffer == nil || startFrame != startingFrame || endFrame != endingFrame || fade.needsUpdate)

Reply:
This seems a little fragile.  I'm sure it works great now, but maintainers need to understand why it's -1.  While clunkier, a notSet Bool would be clearer, hopefully there's a more elegant solution.  In general, I try not to use special values in the place of flags.  An optional might work here too.

Reply:
optional sounds good to me.

Reply:
although it doesn't really matter what the value is as long as it doesn't start with a valid one. The goal is to prevent filling the buffer needlessly